src/catchup/CatchupWork.cpp

```diff
@@ -346,77 +346,46 @@ CatchupWork::onSuccess()
 namespace
 {
 
-// compute first checkpoint that is not 100% finished
-// if lastClosedLedger is not last ledger in checkpoint, then first not finished
-// checkpoint is checkpoint containing lastClosedLedger
-// if lastClosedLedger is last ledger in checkpoint, then first not finished
-// checkpoint is next checkpoint (with 0 ledgers applied)
-uint32_t
-firstNotFinishedCheckpoint(uint32_t lastClosedLedger,
-                           HistoryManager const& historyManager)
-{
-    auto result = historyManager.checkpointContainingLedger(lastClosedLedger);
-    if (lastClosedLedger < result)
-    {
-        return result;
-    }
-    else
-    {
-        return result + historyManager.getCheckpointFrequency();
-    }
-}
-
 // return first ledger that should be applied so at least count history entries
 // are stored in database (count 0 is changed to 1, because even doing only
 // bucket apply gives one entry)
 //
 // if it is impossible, returns smallest possible ledger number -
 // LedgerManager::GENESIS_LEDGER_SEQ
 uint32_t
-firstNeededLedger(CatchupConfiguration const& configuration)
+firstNeededLedger(uint32_t toLedger, uint32_t count)
 {
-    auto neededCount = std::max(configuration.count(), 1u);
-    return configuration.toLedger() > neededCount
-               ? configuration.toLedger() - neededCount + 1
-               : LedgerManager::GENESIS_LEDGER_SEQ;
+    auto neededCount = std::max(count, 1u);
+    return toLedger > neededCount ? toLedger - neededCount + 1
+                                  : LedgerManager::GENESIS_LEDGER_SEQ;
 }
 
-std::pair<bool, uint32_t>
+uint32_t
 computeCatchupStart(uint32_t smallestLedgerToApply,
```

Comment:
Imagine you have CatchupConfiguration set to `{toLedger = 639, count = 100}` and only GENESIS_LEDGER is present in database.

Then:
- smallestLedgerToApply will be 639 - 100 + 1 = 540
- if we make it catchup start, it will result in checkpointRange.first in `CatchupWork::doWork` to be equal 575, which will apply buckets at that checkpoint and only replay ledgers 576..639, which is only 64 ledgers
- if we continue doing what we are now, then smallestCheckpointToApply will be 576 - 64, which will result in 128 ledgers applied, which is better (because we assume that end user wants 100 ledgers in history, which we satisfy now).

Reply:
I see, thanks for unpacking! I think what you described will not happen in the current state if `smallestLedgerToApply` is the last ledger in checkpoint (for example, suppose `toLedger=226, count=100`; `smallestLedgerToApply` would be 127, which will be catchup start, so bucket will be applied at this checkpoint, then ledgers 128 though 226 will be applied, which is 1 ledger short of 100). So I think this statement can be removed:
```
    else if (smallestLedgerToApply == smallestCheckpointToApply)
    {
        return smallestCheckpointToApply;
    }
```

Reply:
good point!